### DISCUSSION TITLE
Handling side-effects of rtk queries

### DISCUSSION THREAD

**Discussion by dbartholomae** [reduxjs/redux-toolkit#3101] [Q&A] [answered]
Hi there!

How to best handle side-effects of rtk queries? E.g. I have a query that checks whether a user is still logged in. When they get logged out, I would like to trigger a popup.
A naive approach could be:

```tsx
function useAuth () {
  const {authData} = useAuthQuery();
  useEffect(() => {
    
  }, [authData]);
}
```

but that's not what `useEffect` is supposed to be used for. Another idea I had was to use a listener:
```ts
listenerMiddleware.startListening({
  predicate: (action) => action.type === 'auth/executeQuery/fulfilled' && action.meta.arg.endpointName === 'authStatus' && !action.payload.loggedIn,
  effect: (_action, listenerApi) => {
    listenerApi.dispatch(showLoggedOutPopUp());
  }
});
```

But I can't actually get the action whose type I'm referencing out of the api, so this looks like it is relying on implementation details.

What's the best approach? If the second one is of interest, would it make sense to expose these actions to be able to react to them? There is already ".internalActions", but this doesn't expose the query action in a way that can be reached, because the action itself is created via a thunk which isn't exposed.
If there is interest in something like this, I'm also happy to provide a PR (with a little guidance).

**Comment by markerikson** [comment, marked as answer]
You can use the endpoint matcher:

```ts
listenerMiddleware.startListening({
  matcher: api.endpoints.myEndpoint.matchFulfilled,
  effect: (action, listenerApi) => {}
})
```

Or if you do have an additional check:

```ts
predicate: (action) => api.endpoints.myEndpoint.matchFulfilled(action) && !action.payload.loggedIn
```

**Reply by dbartholomae** [reply]
Thanks! That's what I was looking for.

**Reply by VictorPulzz** [reply]
That is really cool feature!

**Reply by jdshaeffer** [reply]
Hypothetically, what if the api endpoint you're matching is called in multiple components/views (for whatever reason). Say you want the side effect to run in one component, but not in any other. Is that when you'd have to revert to syncing using `useEffect` or is there a better way to handle that logic at this level?

**Reply by markerikson** [reply]
@jdshaeffer add a listener middleware entry from within the component itself:

- https://redux-toolkit.js.org/api/createListenerMiddleware#adding-listeners-inside-components
